### DISCUSSION TITLE
Can't install on Windows using an Administrator user

### DISCUSSION THREAD

**Discussion by 0xdevalias** [withcatai/node-llama-cpp#353] [General]
This comes from an issue a user opened on the `humanify` repo:

- https://github.com/jehna/humanify/issues/135

And after some debugging/exploration, it seems it is related to `node-llama-cpp`:

- https://github.com/jehna/humanify/issues/135#issuecomment-2381851607

I think it might be a permissions/antivirus/similar sort of a problem, but I haven't used windows as a dev machine in MANY years, so I'm not too sure of what the 'common pitfalls' are with using node on it these days.

I figured I would open a discussion here as well, so that at the very least, if someone else runs into similar with `node-llamma-cpp` in future, at least these errors will show up when searching.

I believe this is the main part of the 'breaking errors':

```
npm error Failed to load a prebuilt binary for platform "win" "x64", falling back to building from source. Error: Error: The specified module could not be found.
npm error \\?\C:\Users\Administrator\AppData\Roaming\npm\node_modules\humanifyjs\node_modules\@node-llama-cpp\win-x64\bins\win-x64\llama-addon.node
npm error     at Module._extensions..node (node:internal/modules/cjs/loader:1586:18)
npm error     at Module.load (node:internal/modules/cjs/loader:1288:32)
npm error     at Module._load (node:internal/modules/cjs/loader:1104:12)
npm error     at Module.require (node:internal/modules/cjs/loader:1311:19)
npm error     at require (node:internal/modules/helpers:179:18)
npm error     at loadBindingModule (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/bindings/getLlama.js:390:25)
npm error     at loadExistingLlamaBinary (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/bindings/getLlama.js:276:37)
npm error     at async getLlamaForOptions (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/bindings/getLlama.js:133:27)
npm error     at async Object.handler (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/cli/commands/OnPostInstallCommand.js:12:13) {
npm error   code: 'ERR_DLOPEN_FAILED'
npm error }
```

Though this part also may help indicate aspects of the issue:

```
npm warn cleanup   [
npm warn cleanup     'C:\\Users\\Administrator\\AppData\\Roaming\\npm\\node_modules\\humanifyjs\\node_modules',
npm warn cleanup     [Error: EPERM: operation not permitted, rmdir 'C:\Users\Administrator\AppData\Roaming\npm\node_modules\humanifyjs\node_modules\isolated-vm\src'] {
npm warn cleanup       errno: -4048,
npm warn cleanup       code: 'EPERM',
npm warn cleanup       syscall: 'rmdir',
npm warn cleanup       path: 'C:\\Users\\Administrator\\AppData\\Roaming\\npm\\node_modules\\humanifyjs\\node_modules\\isolated-vm\\src'
npm warn cleanup     }
npm warn cleanup   ],
npm warn cleanup   [
npm warn cleanup     'C:\\Users\\Administrator\\AppData\\Roaming\\npm\\node_modules\\humanifyjs\\node_modules\\node-llama-cpp',
npm warn cleanup     [Error: EPERM: operation not permitted, rmdir 'C:\Users\Administrator\AppData\Roaming\npm\node_modules\humanifyjs\node_modules\node-llama-cpp\llama\llama.cpp'] {
npm warn cleanup       errno: -4048,
npm warn cleanup       code: 'EPERM',
npm warn cleanup       syscall: 'rmdir',
npm warn cleanup       path: 'C:\\Users\\Administrator\\AppData\\Roaming\\npm\\node_modules\\humanifyjs\\node_modules\\node-llama-cpp\\llama\\llama.cpp'
npm warn cleanup     }
npm warn cleanup   ],
npm warn cleanup   [
npm warn cleanup     'C:\\Users\\Administrator\\AppData\\Roaming\\npm\\node_modules\\humanifyjs',
npm warn cleanup     [Error: EPERM: operation not permitted, rmdir 'C:\Users\Administrator\AppData\Roaming\npm\node_modules\humanifyjs'] {
npm warn cleanup       errno: -4048,
npm warn cleanup       code: 'EPERM',
npm warn cleanup       syscall: 'rmdir',
npm warn cleanup       path: 'C:\\Users\\Administrator\\AppData\\Roaming\\npm\\node_modules\\humanifyjs'
npm warn cleanup     }
npm warn cleanup   ]
npm warn cleanup ]
```

You can see my full deep dive into tracing the code/source of the error/musing on the potential reasons in this comment:

- https://github.com/jehna/humanify/issues/135#issuecomment-2381851607

But this was sort of my summary conclusion based on what I had seen so far:

> ..snip..
> 
> Within that, we can see there is a call to `loadExistingLlamaBinary`:
> 
> - https://github.com/withcatai/node-llama-cpp/blob/4ee10a90e954df26e9e7782d4e2f9e98daab24a5/src/bindings/getLlama.ts#L367-L387
> - https://github.com/withcatai/node-llama-cpp/blob/4ee10a90e954df26e9e7782d4e2f9e98daab24a5/src/bindings/getLlama.ts#L464-L623
> 
> That calls `getPrebuiltBinaryPath`, which then calls `getPrebuiltBinariesPackageDirectoryForBuildOptions`, which is what handles attempting to import the various prebuilt binary packages:
> 
> - https://github.com/withcatai/node-llama-cpp/blob/4ee10a90e954df26e9e7782d4e2f9e98daab24a5/src/bindings/getLlama.ts#L547-L553
> - `getPrebuiltBinaryPath`: https://github.com/withcatai/node-llama-cpp/blob/4ee10a90e954df26e9e7782d4e2f9e98daab24a5/src/bindings/utils/compileLLamaCpp.ts#L286-L313
> - `getPrebuiltBinariesPackageDirectoryForBuildOptions`: https://github.com/withcatai/node-llama-cpp/blob/4ee10a90e954df26e9e7782d4e2f9e98daab24a5/src/bindings/utils/compileLLamaCpp.ts#L364-L426
> 
> ```js
>     } else if (buildOptions.platform === "win") {
>         if (buildOptions.arch === "x64") {
>             if (buildOptions.gpu === "cuda")
>                 // @ts-ignore
>                 return getBinariesPathFromModules(() => import("@node-llama-cpp/win-x64-cuda"));
>             else if (buildOptions.gpu === "vulkan")
>                 // @ts-ignore
>                 return getBinariesPathFromModules(() => import("@node-llama-cpp/win-x64-vulkan"));
>             else if (buildOptions.gpu === false)
>                 // @ts-ignore
>                 return getBinariesPathFromModules(() => import("@node-llama-cpp/win-x64"));
>         } else if (buildOptions.arch === "arm64")
>             // @ts-ignore
>             return getBinariesPathFromModules(() => import("@node-llama-cpp/win-arm64"));
>     }
> ```
> 
> This then calls `getBinariesPathFromModules` which seems to do some stuff to find the bins dir:
> 
> - https://github.com/withcatai/node-llama-cpp/blob/4ee10a90e954df26e9e7782d4e2f9e98daab24a5/src/bindings/utils/compileLLamaCpp.ts#L365-L383
> 
> Jumping back out to `loadExistingLlamaBinary`, we can see the 'failed to load a prebuilt binary' message here within a `catch` block:
> 
> - https://github.com/withcatai/node-llama-cpp/blob/4ee10a90e954df26e9e7782d4e2f9e98daab24a5/src/bindings/getLlama.ts#L598-L611
> 
> The actual triggering error is appended to the end of that message, so the root issue appears to be this part:
> 
> ```
> Error: Error: The specified module could not be found.
> npm error \\?\C:\Users\Administrator\AppData\Roaming\npm\node_modules\humanifyjs\node_modules\@node-llama-cpp\win-x64\bins\win-x64\llama-addon.node
> npm error     at Module._extensions..node (node:internal/modules/cjs/loader:1586:18)
> npm error     at Module.load (node:internal/modules/cjs/loader:1288:32)
> npm error     at Module._load (node:internal/modules/cjs/loader:1104:12)
> npm error     at Module.require (node:internal/modules/cjs/loader:1311:19)
> npm error     at require (node:internal/modules/helpers:179:18)
> npm error     at loadBindingModule (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/bindings/getLlama.js:390:25)
> npm error     at loadExistingLlamaBinary (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/bindings/getLlama.js:276:37)
> npm error     at async getLlamaForOptions (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/bindings/getLlama.js:133:27)
> npm error     at async Object.handler (file:///C:/Users/Administrator/AppData/Roaming/npm/node_modules/humanifyjs/node_modules/node-llama-cpp/dist/cli/commands/OnPostInstallCommand.js:12:13) {
> npm error   code: 'ERR_DLOPEN_FAILED'
> npm error }
> ```
> 
> @Yusufkulcu So I guess the first thing I would be doing is checking whether that file actually exists, and if there is anything that might be blocking it from being able to be loaded (permissions, antivirus, etc):
> 
> - `C:\Users\Administrator\AppData\Roaming\npm\node_modules\humanifyjs\node_modules\@node-llama-cpp\win-x64\bins\win-x64\llama-addon.node`
> 
> If that doesn't help, you may be able to get some more specific ideas/support on the GitHub Discussions page for `node-llama-cpp`:
> 
> - https://github.com/withcatai/node-llama-cpp/discussions/categories/general
> 
> _Originally posted by @0xdevalias in https://github.com/jehna/humanify/issues/135#issuecomment-2381851607_

**Comment by giladgd** [comment]
@0xdevalias I run some tests on the latest version (`3.0.3`), and it seems to work as expected on Windows.
I've tested it on both a Windows 10 machine and on a Windows 11 machine.

From the logs you attached, I think there's something wrong in the installation on the user machine, whether it's npm that hasn't installed the prebuilt binaries properly (maybe they're using some npm mirror that doesn't work as expected), maybe it's related to permissions on the machine itself (like running `npm install` in the folder from one user and trying to use it from another use that doesn't have proper access to the `node_modules` directory), and maybe it has something to do with some program on the machine the blocks the access to the `.node` file.

From the [linked issue](https://github.com/jehna/humanify/issues/135) I can see in the logs that it attempted to fallback to building from source, and since CMake is not installed on the user's machine, it tried to call [`xpm`](https://www.npmjs.com/package/xpm) to install CMake locally inside `node-llama-cpp`, but it failed with this error:
```
npm error ^[[?25l^[[?25h'xpm' is not recognized as an internal or external command,
npm error operable program or batch file.
npm error Failed to build llama.cpp with no GPU support. Error: SpawnError: Command npm exec --yes -- xpm@^0.16.3 install @xpack-dev-tools/cmake@latest --no-save exited with code 1
```

The `^[[?25l^[[?25h'xpm'` part seems to indicate one of the following:
* There's and issue with the installation of nodejs/npm on the user machine.
   > `npm exec --yes` is equivalent to `npx -y`, but `npm exec --yes` works better with other runtimes such as Bun, which is why I used it.
* There's an issue with the escaping done by [`cross-spawn`](https://www.npmjs.com/package/cross-spawn) to run the `npm exec --yes -- xpm@^0.16.3` command. I used [`cross-spawn`](https://www.npmjs.com/package/cross-spawn) because it solved many issues with process spawning on Windows, so I think this is less probable.

What I would suggest to try to fix this issue would be:
* Uninstall all global npm modules, including npm, and reinstalling nodejs, since it seems like there's an issue with its installation.
* Try to use [Bun](https://bun.sh), since it seems they've fixed many Windows compatibility issues.
* [Install CMake](https://cmake.org/download/) manually, so `node-llama-cpp` can use it.

Before doing all of the above, it would be helpful if you can check if `C:\Users\Administrator\AppData\Roaming\npm\node_modules\humanifyjs\node_modules\@node-llama-cpp\win-x64\bins\win-x64\llama-addon.node` exists.

**Reply by giladgd** [reply]
Actually, this may indeed be the permission issue I mentioned.
The attached error shows that `node-llama-cpp` is installed under the `Administrator` user, which should never be done, and is very prone to permission issues.

**Reply by 0xdevalias** [reply]
@giladgd Thanks for testing + your notes. I'm unable to test these directly myself as it was another user who originally had the issue, but I have copied your notes above back onto the original thread, so hopefully we'll hear back from them about it:

- https://github.com/jehna/humanify/issues/135#issuecomment-2412603357
